plugins/connection/paramiko_ssh.py

```diff
@@ -0,0 +1,682 @@
+
+        ssh_connect_kwargs = self._parse_proxy_command(port)
+
+        ssh.set_missing_host_key_policy(MyAddPolicy(self))
+
+        conn_password = self.get_option('password')
+
+        allow_agent = True
+
+        if conn_password is not None:
+            allow_agent = False
+
+        try:
+            key_filename = None
+            if self.get_option('private_key_file'):
+                key_filename = os.path.expanduser(self.get_option('private_key_file'))
+
+            ssh_connect_kwargs['auth_timeout'] = self.get_option('timeout')
+
+            ssh_connect_kwargs['banner_timeout'] = self.get_option('banner_timeout')
+
+            if LooseVersion(paramiko.__version__) >= LooseVersion('2.6.0'):
+                ssh_connect_kwargs['disabled_algorithms'] = disabled_algorithms
+
+            ssh.connect(
+                self.get_option('remote_addr').lower(),
+                username=self.get_option('remote_user'),
+                allow_agent=allow_agent,
+                look_for_keys=self.get_option('look_for_keys'),
+                key_filename=key_filename,
+                password=conn_password,
+                timeout=self.get_option('timeout'),
+                port=port,
+                **ssh_connect_kwargs,
+            )
+        except paramiko.ssh_exception.BadHostKeyException as e:
+            raise AnsibleConnectionFailure(f'host key mismatch for {e.hostname}') from e
+        except paramiko.ssh_exception.AuthenticationException as ex:
+            raise AnsibleAuthenticationFailure() from ex
+        except Exception as ex:
+            msg = str(ex)
+            if "PID check failed" in msg:
+                raise AnsibleError("paramiko version issue, please upgrade paramiko on the machine running ansible") from ex
+            elif "Private key file is encrypted" in msg:
+                msg = (
+                    f"ssh {self.get_option('remote_user')}@{self.get_options('remote_addr')}:{port}"
+                    "\nTo connect as a different user, use -u <username>."
+                )
+                raise AnsibleConnectionFailure(msg) from ex
+            else:
+                raise AnsibleConnectionFailure(msg) from ex
+
+        return ssh
+
+    def exec_command(self, cmd: str, in_data: bytes | None = None, sudoable: bool = True) -> tuple[int, bytes, bytes]:
+        """ run a command on the remote host """
+
+        super().exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        if in_data:
+            raise AnsibleError("Internal Error: this module does not support optimized module pipelining")
+
+        bufsize = 4096
+
+        try:
+            self.ssh.get_transport().set_keepalive(5)
+            chan = self.ssh.get_transport().open_session()
+        except Exception as e:
+            text_e = to_text(e)
+            msg = "Failed to open session"
+            if text_e:
+                msg += f": {text_e}"
+            raise AnsibleConnectionFailure(to_native(msg)) from e
```

Comment:
And this `to_native()` here is redundant either way.
```suggestion
            raise AnsibleConnectionFailure(msg) from e
```